### PR TITLE
[Function Arguments]: Add basic functions for Issue #2354 

### PR DESCRIPTION
**Pull request by asrvd**
@BethanyG 
For issue #2354 

I have started to work on the issue as I managed to get some free time, I have put some basic functions in, please review them and let me know about changes needed.

### CODE REVIEW

**Comment by BethanyG** [comment]
@asheeeshh  - awesome news!  I am a bit underwater with work at the moment, but will make sure to review & get you feedback by the end of the weekend.  Apologies for the delay.

File: exercises/concept/himaris-cake-shop/.meta/exemplar.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# For 1kg cake
```
Comment (BethanyG) [inline exercises/concept/himaris-cake-shop/.meta/exemplar.py:1]:
```suggestion
# For aprox. 1kg - 1.3kg cake - an 20cm/8in round.  Serves 8-20 people.
```

File: exercises/concept/himaris-cake-shop/.meta/exemplar.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    for price in ingredients_prices:
+        PRICE += price
+    return PRICE
+
+def calculate_total_price(weight_of_base=1, number_of_layers=1, **ingredients_prices): # *kwargs method
+    return (calculate_base_price(weight_of_base) * number_of_layers) + total_ingredients_price(*ingredients_prices.values())
+
+def make_cake():
+    # Don't know what to put in here yet or whether we need this
+    pass
```
Comment (asheeeshh) [inline exercises/concept/himaris-cake-shop/.meta/exemplar.py:22]: any ideas to implement this? @BethanyG 
and I also wanted to know if the sequencing of functions is correct according to sequence of learning, do we need any addition her?

Comment (BethanyG) [reply exercises/concept/himaris-cake-shop/.meta/exemplar.py:22]: Hi @asheeeshh 

Sorry for the time it's taken to get back to you.  😞   I was playing with this this weekend.  I think it's a good _start_, but the scenario felt too simple to me, so I was trying to draft something a little more involved/detailed.  Something that would demonstrate _why_ you would pass arbitrary args in one form or another. But I might have made a mess of it.   I do think that the order we present the types is correct:

 1. positional
 2. positional + named/default, 
 3. positional + default + arbitrary
 4. positional + default + arbitrarily named
 5. _Didn't come up with something for this (yet!)_

I am going to spend some more time today cleaning up what I did this weekend, and when I am done I will post it here to see what you think.  I might also come back and adjust that pricing list (_should we do it in Yen??_).

So far, I have the following story and "pricing", but am trying to work out how those things would now map to tasks:

## Story

Himari is opening a new custom cake shop, and has asked for your help with her costing, online ordering, and POS (Point of Sale) systems.

She is really into flavors, fillings, and decorations, so has decided to focus on just two sizes: 15cm/6in and 20cm/8in rounds. 

Cakes are priced by their **weight** (_starting with a base weight of aprox 1kg, and going up by 1/2kg for each layer added_). Each 20cm layer is approximately 1/2kg (_she divides the thicker single cake in half._). Each 15cm layer is _also_ 1/2kg (_15cm layers are thicker than 20cm layers_). 

There are additional charges for each custom frosting, filling, or topping that is added to a cake order. There is also a surcharge for bakes that take a longer amount of time (_cakes over 4 layers_) to cook or to decorate.

Each basic 2-layer cake is filled with vanilla custard and frosted with fresh whipped cream.  While she does offer a 1kg single-layer 20cm cake, it comes with the filling or custard on the side - so there is no filling discount for a single-layer cake. The pricing for base filling and frosting is included in the price for each added tier.

**Note:** Himari's oven can fit up to 6 15cm/6in cake pans and 4 20cm/8in cake pans at a time, and making decorations and various fillings and frostings in bulk also saves time --  but we're keeping bake times and decoration times linear for simplicity of calculation.

Himari's pricing and timing tables are below:

| **Item**                                                                                                                                                	| **Price** 	|
|---------------------------------------------------------------------------------------------------------------------------------------------------------	|-----------	|
| Small Size Discount (_for customizing cakes that are 15cm_)                                                                                             	| 0.85      	|
| Base Price (_1kg - 1.3kg of cake_)<br> This is a 1 or 2 layer 20cm cake or a 2 layer 15cm cake.                                                                                                                    	| 200       	|
| Tier Price (_.5kg - .65kg of cake_)<br> This is 1 layer of a 20cm or 15cm cake.                                                                                                                     	| 170       	|
| Customization Surcharge                                                                                                                                 	| 5         	|
| Chocolate Ganache Frosting                                                                                                                              	| 55        	|
| Flavored Buttercream Frosting (_vegan and vegetarian available_)<br>Vanilla, Chocolate, Raspberry, Blackberry, Grapefruit, Orange, Lemon, Coconut, Lime 	| 50        	|
| Flavored Whipped Cream Frosting (_vegan & vegetarian available_)<br>Vanilla, Chocolate, Raspberry, Blackberry, Grapefruit, Orange, Lemon, Coconut, Lime 	| 40        	|
| Fruit Glaze                                                                                                                                             	| 30        	|
| Honey Glaze                                                                                                                                             	| 30        	|
| Flavored Custard (_vegan & vegetarian available_)<br>Coconut, Hazelnut, Pistachio, or Sesame                                                            	| 30        	|
| Chocolate Ganache Filling                                                                                                                               	| 50        	|
| Nuts Mixin <br>Coconut, Hazelnut, Pistachio, or Sesame                                                                                                  	| 40        	|
| Fruit Mixin <br>Strawberries, Blackberries, Raspberries, Gooseberries, Grapes, Melon, or Citrus                                                         	| 50        	|
| Fruit Filling (_replaces custard with fruit_)<br>Strawberries, Blackberries, Raspberries, Gooseberries, Grapes, Melon, or Citrus                        	| 5         	|
| Jam Filling (_replaces custard with Jam_)<br>Strawberry, Blackberry, Raspberry, Marmalade, or Grape                                                     	| 5         	|
| Berry Topping <br>Strawberries, Blackberries, Raspberries, Gooseberries                                                                                 	| 50        	|
| Sliced Fruit Topping <br>Mangoes, Oranges, Melon, Grapes, Kiwi, Banana                                                                                  	| 40        	|
| Mixed Fruit Topping                                                                                                                                     	| 30        	|
| Nut Topping <br>Coconut, Hazelnut, Pistachio, or Sesame                                                                                                 	| 25        	|

<br><br>

| **Item**              	| **Time to Make** 	|
|-----------------------	|------------------	|
| Time to Bake Base     	| 40               	|
| Time to Bake Tier     	| 20               	|
| Time to Decorate Base 	| 15               	|
| Time to Decorate Tier 	| 5                	|

**Comment by BethanyG** [comment]
Here's what I came up with for functions.  Still trying to think through how we'd handing the `positional-only` vs `keyword-only` stuff.  Will post again if I think of something.  Maybe producing a customer receipt with itemized and labeled items?  🤔 

```python
PRICING = {
"small_size_discount": .85,
"base_price": 200,
"tier_price": 170,
"time_to_bake_base": 40,
"time_to_bake_tier": 20,
"base_time_to_decorate": 15,
"tier_time_to_decorate": 5,
"customization_surcharge": 5,
"Chocolate Ganache Frosting": 55,
"Flavored Buttercream": 50,
"Flavored Whipped Cream": 40,
"Fruit Glaze": 30,
"Honey Glaze": 30,
"Flavored Custard": 30,
"Chocolate Ganache Filling": 50,
"Nuts Mixin": 40,
"Fruit Mixin": 50,
"Fruit Filling": 5,
"Jam Filling": 5,
"Berry Topping": 50,
"Sliced Fruit Topping": 40,
"Mixed Fruit Topping": 30,
"Nut Topping": 25
}


order= {"size": "15cm",  
   "Flavored Buttercream": "Blackberry",
   "Flavored Custard": "Sesame",
   "Fruit Mixin": "Melon",
   "Berry Topping": "Blackberry"
   }

def base_price(number_of_cakes, added_layers):

    return (PRICING["base_price"] * number_of_cakes) +  (PRICING["tier_price"] * added_layers)


# Positional with named/default values.
def time_for_baking_and_decorating(number_of_cakes, added_layers=0):
   base_time = (number_of_cakes * PRICING["time_to_bake_base"]) + (number_of_cakes * PRICING["base_time_to_decorate"])
   tier_time = (number_of_layers * PRICING["time_to_bake_tier"]) + (number_of_layers * PRICING["tier_time_to_decorate"])

   return base_time + tier_time

# Positonal with arbitrary addon ingredients
def customization_price(cake_size="20cm", *addons):
    added_price = sum(*addons)
    
    return added_price * PRICING["small_size_discount"] if cake_size=="15cm" else added_price

# Positional with named(default) and arbitrary order dict.
def total_order_price(number_of_cakes, added_layers=0,  **kwargs): # **kwargs method
    base_cost = base_price(number_of_cakes, added_layers)
    customizations = tuple((PRICING[item] for item in kwargs.keys() if item !="size"))
    customization_cost = customization_price(kwargs["size"], customizations)
    surcharge = 0 if added_layers < 2 else PRICING["customization_surcharge"]
    
    return base_cost + customization_cost + surcharge

print(total_order_price(3, added_layers=4, **order))
>>>1438.0
```

**Comment by BethanyG** [comment]
Please feel free to adapt any of this, or ignore it if it doesn't fit with your ideas.  I do think that you are on the right track.  😄   I'll see what I can come up with for the 5th task.

**Comment by BethanyG** [comment]
@asheeeshh - just checking in -- are you still wanting to work on this?  Just let me know.  Thanks!

**Comment by asheeeshh** [comment]
> @asheeeshh - just checking in -- are you still wanting to work on this? Just let me know. Thanks!

sorry for the late reply, but currently I'm a little too preoccupied with exams and stuff so I'm barely working on anything other than some personal projects, but I definitely want to complete this exercise, if I could be given some time I'll surely get back to it once I'm done with exams probably in last week of june.

**Comment by asrvd** [comment]
@BethanyG I'm not quite sure how we would implement positional only parameters, but keyword only parameters can be used to maybe display a list of all the extras like toppings which are added 🤔

**Pull request closed without merging**